python/GafferSceneUITest/SceneUITestCase.py

```diff
@@ -0,0 +1,120 @@
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import math
+import unittest
+
+import imath
+
+import IECore
+import IECoreGL
+
+import Gaffer
+import GafferUI
+import GafferScene
+import GafferTest
+
+import GafferUITest
+
+class SceneUITestCase( GafferUITest.TestCase ) :
+
+	# Sets up a grid of instances on the XY plane of the supplied source
+	# prototype scene. Each instance will have a random rotation applied. The
+	# supplied ScenePlug's node needs to be parented to a Gaffer.ScriptNode.
+	@staticmethod
+	def setupInstancer( prototypeScenePlug, copies = 2500, spacing = 5 ) :
+
+		divisions = math.ceil( math.sqrt( copies ) )
+		dimension = divisions * spacing
+		assert( divisions > 0 )
+
+		script = prototypeScenePlug.node().ancestor( Gaffer.ScriptNode )
+		script["_Plane"] = GafferScene.Plane()
+		script["_Instancer"] = GafferScene.Instancer()
+		script["_PathFilter"] = GafferScene.PathFilter()
+		script["_Instancer"]["in"].setInput( script["_Plane"]["out"] )
+		script["_Instancer"]["prototypes"].setInput( prototypeScenePlug )
+		script["_Instancer"]["filter"].setInput( script["_PathFilter"]["out"] )
+		script["_Plane"]["dimensions"].setValue( imath.V2f( dimension ) )
+		script["_Plane"]["divisions"].setValue( imath.V2i( divisions ) )
+		script["_PathFilter"]["paths"].setValue( IECore.StringVectorData( [ "/plane" ] ) )
+		script["_Transform"] = GafferScene.Transform( "Transform" )
+		script["_Transform"]["in"].setInput( script["_Instancer"]["out"] )
+		script["_Random"] = Gaffer.Random( "Random" )
+		script["_Expression"] = Gaffer.Expression( "Expression" )
+		script["_PathFilter1"] = GafferScene.PathFilter( "PathFilter1" )
+		script["_Transform"]["filter"].setInput( script["_PathFilter1"]["out"] )
+		script["_Random"]["contextEntry"].setValue( 'scene:path' )
+		script["_Random"]["baseColor"].setValue( imath.Color3f( 0.5, 0, 0 ) )
+		script["_Random"]["hue"].setValue( 1.0 )
+		script["_Random"]["saturation"].setValue( 1.0 )
+		script["_Random"]["value"].setValue( 1.0 )
+		script["_PathFilter1"]["paths"].setValue( IECore.StringVectorData( [ '/*/instances/*/*' ] ) )
+		script["_Expression"].setExpression( 'import imath\n\nr = parent["_Random"]["outColor"]\nr *= 360\n\nparent["_Transform"]["transform"]["rotate"] = r', "python" )
+
+		return script["_Transform" ]["out"]
+
+	# Renders the supplied scene a fixed number of times using the
+	# GafferTest.TestRunner.PerformanceScipe to record timings.
+	def benchmarkRender( self, scenePlug, frames = 500, rendererName = "OpenGL" ) :
+
+		window = GafferUI.Window()
+		gl = GafferUI.GLWidget()
+		window.setChild( gl )
+
+		gl._qtWidget().setMinimumSize( 512, 512 )
+		gl._qtWidget().setMaximumSize( 512, 512 )
+
+		window.setVisible( True )
+		gl._makeCurrent()
```

Comment:
I think there's something not quite sufficient about the way we're managing GL here. Sometimes I see a solid yellow window, indicating that maybe the camera is badly placed. Sometimes I see the content, but only occupying the top-left corner of the window, indicating that maybe our GL viewport is wrong. Not sure if we're maybe in an awkward half-way position between using GLWidget the way it intends, and just managing a context of our own.